core/src/main/kotlin/in/specmatic/core/pattern/JSONObjectPattern.kt

```diff
@@ -102,9 +103,27 @@ data class JSONObjectPattern(override val pattern: Map<String, Pattern> = emptyM
 
 fun generate(jsonPattern: Map<String, Pattern>, resolver: Resolver): Map<String, Value> {
     val resolverWithNullType = withNullPattern(resolver)
-    return jsonPattern.mapKeys { entry -> withoutOptionality(entry.key) }.mapValues { (key, pattern) ->
-        attempt(breadCrumb = key) { resolverWithNullType.generate(key, pattern) }
-    }
+
+    val optionalProps = jsonPattern.keys.filter { isOptional(it) }.map { withoutOptionality(it) }
+
+    return jsonPattern
+        .mapKeys { entry -> withoutOptionality(entry.key) }
+        .mapValues { (key, pattern) ->
+            attempt(breadCrumb = key) {
+                try {
+                    Optional.of(resolverWithNullType.generate(key, pattern))
+                } catch (e: PatternCycleException) {
+                    if (optionalProps.contains(key)) {
```

Comment:
@westse In this location, the optionality of the key itself may be relevant to break the cycle. What we should checking for is if the nullability of the value. We will push this fix to your branch shortly so that you can also take a look at the changes and share your thoughts. Hope that is okay. cc @joelrosario

Reply:
Sorry, I'm not sure I follow. I may be missing a concept, but let me describe the problem I'm addressing and this current solution. (Also, I apologize that somehow I originally missed including a unit test to expose the limitation being addressed. I have since pushed a test along with a necessary refactor.)

Stepping back, this PR is meant to have specmatic behave well in the face of cycles in patterns (e.g. A->B->A). Per @joelrosario 's feedback, cycles in required, non-nullable patterns should result in an exception. However cycles in nullable patterns (both required and optional) can be allowed because we can rely on AnyPattern.generate()'s behavior to eventually randomly select null and terminate the cycle. (That is what the Resolver cyclePreventionStack was for)

However, there is another case: cycles where one or more properties are optional (ie. do no appear in the required list). Rather than failing due to a cycle it seems preferrable (and contract-valid) to instead omit the optional propery in the generated stub. That is what this logic is meant to accomplish.